### DISCUSSION TITLE
RP Pico W BLE to USB MIDI

### DISCUSSION THREAD

**Discussion by gonzob** [tttapa/Control-Surface#1070] [General]
Hi.
I'm trying to set up a PicoW to receive BLE MIDI from an instrument, and send it to USB-MIDI.

Using the ArduinoIDE, I set the USB stack to "Adafruit TinyUSB", and the IP/Bluetooth stack to "IPV4+IPV6+Bluetooth-32K". It all compiles and loads OK. I can see the USB MIDI device plugged into my PC, and (using my phone) I can see the BLE device, but it is set up as a peripheral, not a central. 
How to I get the BLE to be a central so it can connect to my peripheral instruments?

When I connect to the BLE device with my phone, the LED comes on - so I think that's working.

Code:

`#include <Control_Surface.h>
#include <MIDI_Interfaces/BluetoothMIDI_Interface.hpp>

// Instantiate a MIDI over USB interface
USBMIDI_Interface midi_usb;

// Instantiate a MIDI over BLE interface
BluetoothMIDI_Interface midi_ble;

// Instantiate the pipe to connect the two interfaces
BidirectionalMIDI_Pipe pipes;
 
void setup() {
  // Change the name of the BLE device (must be done before initializing it)
  midi_ble.setName("BLE-PICO");
  // Manually route MIDI input from the BLE interface to the USB interface,
  // and the MIDI input from the USB interface to the BLE interface
  midi_ble | pipes | midi_usb;
  // Initialize the MIDI interfaces
  MIDI_Interface::beginAll();
  // Initialize the built-in LED
  pinMode(LED_BUILTIN, OUTPUT);

}
 
void loop() {
  // Continuously poll all interfaces and route the traffic between them
  MIDI_Interface::updateAll();

  // Display the connection status using the built-in LED
  digitalWrite(LED_BUILTIN, midi_ble.isConnected() ? HIGH : LOW);

}
`

**Comment by tttapa** [comment]
Central mode is currently not supported. If you have some BLE background, adding it to your application should be _relatively_ straightforward though, as most of the Control Surface components can be reused.

Here's the rough outline:

1. Create a copy of the `BTstackBackgroundBackend` class, let's call it `BTstackBackgroundBackendCentral`.
2. In `BTstackBackgroundBackendCentral`, replace the calls to `midi_ble_btstack::{init,notify}` by calls to new functions in a new namespace, e.g. `midi_ble_btstack_central::{init,notify}`.
3. Create a MIDI interface out of your `BTstackBackgroundBackendCentral`, for example, use `struct BluetoothCentralMIDI_Interface : GenericBLEMIDI_Interface<BTstackBackgroundBackendCentral> {};`.
4. Now for the harder part: look at the [sm_pairing_central.c](https://github.com/bluekitchen/btstack/blob/master/example/sm_pairing_central.c) example in btstack. Adapt it to your needs so it connects to your MIDI peripheral, and take care of bonding, secure connections, configuring connection parameters, discovering services and characteristics, etc. This should be standard BLE stuff.
5. If you were able to successfully connect to a peripheral and discovered its MIDI service, implement the `midi_ble_btstack_central::{init,notify}` functions we introduced earlier. The `init` function initiates the connection, and sets up some callbacks (see the `MIDIBLEInstance` interface in `BLEAPI.hpp`), and the `notify` function sends a single BLE packet to the peripheral (note: if the Arduino is a peripheral, it should use a notification, if it's a central, it should use a standard write/no-response instead). You can have a look at https://github.com/tttapa/Control-Surface/blob/main/src/MIDI_Interfaces/BLEMIDI/BTstack/gatt_midi.cpp for inspiration.

**Comment by gonzob** [comment]
Thanks Pieter.

I think it's beyond my current skills.

Gonzo.
